### PR TITLE
Deprecate package object with extends clause

### LINKED ISSUE #441: Drop package objects with inheritance
Replace with support for top-level definitions?

### PR DESCRIPTION
**Pull request by adriaanm**
**UPDATE:** the deprecation ended up being put behind `-Xsource:2.14` (see https://github.com/scala/scala/pull/7857)

Fix scala/scala-dev#441

Workaround: use a regular object and import it.  I consider the explicit import an advantage: it makes it more clear where implicits are coming from. In general, I would not recommend defining implicits in package objects.

Longer term, package objects will be replaced by top-level definitions.

To enable this migration to top-level definitions as a replacement
for package objects, we must first remove the main aspect of
package objects that complicates their implementation and
that would not longer be supported by top-level definitions.

Once that's done, the migration would mean rewriting
`package object foo { ... }` to `package foo { ... }`,
assuming the `...` are exclusively definitions, and not statements.

Losing inheritance for package objects means that some code duplication
will be required (you'll have to add val/type aliases for members,
instead of inheriting the members directly). You also can't use inheritance
to prioritize implicits, but, implicits defined in packages are also
being phased out (i.e., package prefixe don't contribute to the implicit scope).

The main complication with supporting full-on package objects is
that name resolution needs to know all definitions in a package
in order to resolve the parents of said package, and when there's
a cycle there (a package object inherits a class defined in the package),
the implementation becomes challenging due to the open nature of packages.

While we can easily rule out the cycle in the following snippet that
uses plain objects, with package objects it's not so easy:

```
object Dogs extends Dogs.Animal { // cycle!
  class Animal
}
```

The spec currently has similar awkwardness in place to deal with this:

> The package object should not define a member with the same name as
> one of the top-level objects or classes defined in package $p$. If
> there is a name conflict, the behavior of the program is currently
> undefined. It is expected that this restriction will be lifted in a
> future version of Scala.

Finally, package objects are objects in name only -- they already are
not first-class values, and now they also can't extend anything. This
leaves them as simply syntactic baggage. Top-level definitions are
intuitive, require no syntax, and can be implemented in the same way
as package objects behind the scenes, albeit without the issue of cycles.

```
package object foo

class D {
  println(foo) // error: package foo is not a value
}
```

### CODE REVIEW

**Comment by smarter** [comment]
Based on the discussion in https://github.com/scala/scala-dev/issues/441 this seems to be a pretty controversial change, are you intending to go ahead with this anyway ?

**Comment by adriaanm** [comment]
Yep

**Comment by adriaanm** [comment]
Seriously, though, I read that discussion as having valid pros and cons for deprecation, though with the pros outnumbering the cons.

**Review by smarter — COMMENTED** [review]
>  I read that discussion as having valid pros and cons for deprecation, though with the pros outnumbering the cons.

Please list these pros/cons in the commit message :).

**Comment by adriaanm** [comment]
will do -- in the mean time, here's a beauty: 357905c

**Comment by adriaanm** [comment]
tests to fix for tickets 
  - scala/bug#5954
  - scala/bug#1987
  - scala/bug#5604

**Comment by dwijnand** [comment]
How does

```scala
object `package` extends C
```

behave, instead?

**Comment by som-snytt** [comment]
The best part is that it blows the future wide open to re-introduce package objects as a module system.

File: src/compiler/scala/tools/nsc/ast/parser/Parsers.scala [hunk 1]
```diff
@@ -2937,13 +2937,13 @@ self =>
...
      */
-    def objectDef(start: Offset, mods: Modifiers): ModuleDef = {
+    def objectDef(start: Offset, mods: Modifiers, isPackageObject: Boolean = false): ModuleDef = {
       in.nextToken()
       val nameOffset = in.offset
       val name = ident()
       val tstart = in.offset
       atPos(start, if (name == nme.ERROR) start else nameOffset) {
-        val template = templateOpt(mods, name, NoMods, Nil, tstart)
+        val template = templateOpt(mods, if  (isPackageObject) nme.PACKAGEkw else name, NoMods, Nil, tstart)
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/ast/parser/Parsers.scala]: scalafmt where are you? SIP-12 where are you? There's always either too many spaces or not enough spaces around a paren. `if ( isPO )` ofc. Also ofc we need API for `name.orElsePackageIf(b)`. But I'd be satisfied with deleting one space, if my new computer glasses worked.

File: src/compiler/scala/tools/nsc/ast/parser/Parsers.scala [hunk 2]
```diff
@@ -3081,6 +3073,9 @@ self =>
       val parentPos = o2p(in.offset)
       val tstart1 = if (body.isEmpty && in.lastOffset < tstart) in.lastOffset else tstart
 
+      // we can't easily check this later, because `gen.mkParents` adds the default AnyRef parent, and we need to warn based on what the user wrote
+      if (name == nme.PACKAGEkw && parents.nonEmpty) deprecationWarning(tstart, s"package object inheritance is deprecated", "2.13.0")
+
```
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/ast/parser/Parsers.scala:3080]: It's a pity it's quite pithy. I wonder if it should say another word, like, "And we're taking package objects away from you entirely in the next release." First they took away my inheritance, then my objects, next there will be no packages at all under the tree. No fun living at the orphanage.

**Review by som-snytt — APPROVED** [review]
OK. People want to take away our cycles, but where would we be without cycle sheds?

File: src/compiler/scala/tools/nsc/ast/parser/Parsers.scala (hunk 2 above)
Comment (adriaanm) [reply src/compiler/scala/tools/nsc/ast/parser/Parsers.scala:3080]: I blame the boomers. Anyway, I added some links and verbiage.

File: src/compiler/scala/tools/nsc/ast/parser/Parsers.scala (hunk 1 above)
Comment (adriaanm) [reply src/compiler/scala/tools/nsc/ast/parser/Parsers.scala]: but! who doesn't like more`          `space! (hint: markdown)

**Comment by neko-kai** [comment]
It's a big change that will break a lot of libraries. Shouldn't this go through a SIP?

**Comment by adriaanm** [comment]
The actual change will. Deprecation won't.

**Comment by neko-kai** [comment]
It's nearly the same thing, though – I doubt it would be 'undeprecated' if the SIP doesn't go through. Seems like a very one-sided decision to me.

**Comment by soronpo** [comment]
>It's nearly the same thing, though – I doubt it would be 'undeprecated' if the SIP doesn't go through. Seems like a very one-sided decision to me.

I tend to agree. A mitigation can be if there was a warning suppression mechanism for the library to use so it can use fatal warnings. Deprecating is a way to say "please don't do this/this is not recommended". That is fine, but we need away to reply to the compiler "I agree, but shut up please".

**Comment by adriaanm** [comment]
Sure, I started working on that here 
https://github.com/adriaanm/scala/tree/deprconf. Help finishing the job very welcome

**Comment by soronpo** [comment]
@adriaanm Can you please schedule a 2.13 community-build for this PR?

**Comment by adriaanm** [comment]
Since it's merged, the community build will pick it up. I assume you're worried about its effect on warnings, but those won't show up in the community build because we already disable `-Xfatal-warnings` there in most projects.

**Comment by godenji** [comment]
@adriaanm from your reply on the [corresponding issue](https://github.com/scala/scala-dev/issues/441#issuecomment-455507729)

> I think it's actually desirable to have a least an explicit entry point (an import in this case) to understanding where implicits come from

But it's not just implicits, you may have non-implicit methods, vals, objects, etc. that you want in scope at the package level. Having to convert `package object extends ...` to plain `object extends ...` and then import that in *every source file* is honestly a massive pain compared to the wonderous status quo that obviates all this (IMO) needless extra work.

This is going to wreak serious havok, every code base built with `-Xfatal-warnings` is going to blow up. What the heck, this is just merged into 2.13, no SIP??

**Comment by dwijnand** [comment]
They can drop fatal warnings or resolve the deprecation warnings.  The compiler development can't be bound by the constraints of fatal warnings.

**Comment by som-snytt** [comment]
Besides using `-Yimports` to add an "auto-imported predef", I'm following up Adriaan's "filter ALL the things" idea at https://github.com/scala/scala/pull/7728, so users can keep fatal warnings by either moving a package object to root import position (which might work well for top packages of a library or app) or just whitelisting the warnings.

Note that `-Xfatal-warnings` will [become](https://github.com/scala/scala/pull/7686) `-Werror`, but the old name will still work without itself emitting a deprecation.

(Edit: now I wish it were called, `-Xmortal-admonitions`.)

**Comment by godenji** [comment]
> They can drop fatal warnings or resolve the deprecation warnings. The compiler development can't be bound by the constraints of fatal warnings.

Well, the issue is that there's no viable replacement. Top-level definitions aren't yet supported; custom predef is globally scoped (IDE support is probably amazing); plain objects require an import in every source file.

I can appreciate that this change will make life easier for the compiler team, but it's strictly worse for users that have benefited for years from `package scoped` auto-imports.

**Comment by dwijnand** [comment]
Another alternative is writing the forwarders in the package object.  But I agree with you those are the tradeoffs for users making use of this feature which is being deprecated.

**Comment by SethTisue** [comment]
@godenji thanks for speaking up with your concerns about this, I think the questions you are asking and the points you're bringing up will be of concern to others as well.

> no SIP??

well, I think everyone on the SIP committee is aware of these changes, and IIRC this has come up in several meetings...

...but on the other hand, I was a bit surprised not to find top level definitions on the list at https://github.com/lampepfl/dotty/issues/5221, and I've now said so at https://github.com/lampepfl/dotty/issues/5221#issuecomment-465736990

> every code base built with `-Xfatal-warnings` is going to blow up.

Adriaan and I both agree with @dwijnand that compiler development and the progress of the language can't be bound by the constraints of fatal warnings.

https://github.com/ghik/silencer provides a escape hatch. another possible escape hatch is to enable fatal warnings only on the main Scala version you are targeting (e.g., 2.12)

trying to cross-compile to multiple Scala versions without any warnings at all in any of the versions you are targeting may sometimes be possible for some projects, but making it possible in general isn't really a goal for us, it's far too constraining.

> Well, the issue is that there's no viable replacement. Top-level definitions aren't yet supported

good point. @adriaanm is there a reason this deprecation couldn't wait for 2.14? are we deprecating in 2.13 because we want to remove it in 2.14 because there's something in the 2.14 roadmap that conflicts?I'd like to understand this as well.

**Comment by SethTisue** [comment]
> I think everyone on the SIP committee is aware of these changes, and IIRC this has come up in several meetings...

ah, just found https://github.com/scala/scala-dev/issues/441#issuecomment-371990261 where I wrote:

> this was discussed by the SIP committee at the SIP meeting this week

unfortunately it seems that the minutes from that meeting were never published; at least, they're missing from https://github.com/scala/docs.scala-lang/tree/master/_sips/minutes (probably a casualty of the changeover in the SIP process lead position, I suppose), but there is a video at https://www.youtube.com/watch?v=dFEkS71rbW8 and the agenda includes "Remove package objects", with a link to https://github.com/scala/scala-dev/issues/441

so essentially https://github.com/scala/scala-dev/issues/441 is the SIP, so it was up for public discussion, and the committee was made aware as well

so process-wise, I think once this is added to the list at 
https://github.com/lampepfl/dotty/issues/5221 we can say the process was followed. _especially_ since this is, for now, only a deprecation

**Comment by adriaanm** [comment]
Top level definitions provide no means for inheritance, which is why we’re
deprecating this aspect of package objects. That’s why we’re deprecating
now so that we can hopefully add top level definitions in 2.14

**Comment by smarter** [comment]
Dotty right now has both package objects with inheritance and top-level objects, so from what I can see there's no urgency to deprecate this.

**Comment by Ichoran** [comment]
There doesn't seem to be any logical reason why package objects' inheritance has to be deprecated in order for top-level definitions to go in in 2.14--or now, for that matter.  Even if you use the package object mechanism for TLDs, you just need to mangle names slightly differently and the two can coexist.  Yes, there's weird stuff about how the two can look into each other, but you just make some ad-hoc rules to simplify the logic during the period where both coexist.  (E.g. allow the visibility to go only one way.)

That the deprecation is going in before the replacement is ready is rather disappointing.  That the envisioned replacement isn't actually a full replacement because it doesn't cleanly solve the implicit priority thing (as far as I can tell) is also disappointing.

I understand the inherent problems that package objects have, and look forward to them going away.  I haven't liked to have to use that mechanism, and I haven't enjoyed the bugs.  But they're here now and work and aid usability.  So without a good alternative for inheritance use cases, the deprecation doesn't feel ready to me.

**Comment by neko-kai** [comment]
I, for one, am NOT looking forward to package objects going away. The fact that top-level is first-class and composable by mixins like any other object is one of the beauties of Scala to me.

**Comment by dwijnand** [comment]
> Even if you use the package object mechanism, you just need to mangle names slightly differently and the two can coexist. Yes, there's weird stuff about how the two can look into each other, but you just make some ad-hoc rules to simplify the logic during the period where both coexist. (E.g. allow the visibility to go only one way.)

There's lots of wonderfully written detail about that "weird stuff" in https://github.com/gkossakowski/kentuckymule/issues/6#issuecomment-378492570.

**Comment by SethTisue** [comment]
> I, for one, am NOT looking forward to package objects going away

please use https://github.com/scala/scala-dev/issues/441 for any general discussion, so we can keep the discussion here on the pull request focused on 1) the implementation, and 2) the timing of the deprecation

**Comment by adriaanm** [comment]
Again, please note that we’re only talking about deprecating the EXTENDS
clause of a package object. Not package objects themselves. Once your code
base is free of  package objects that use inheritance, migrating to
toplevel definitions is trivial.

**Comment by soronpo** [comment]
> 
> 
> Since it's merged, the community build will pick it up. I assume you're worried about its effect on warnings, but those won't show up in the community build because we already disable `-Xfatal-warnings` there in most projects.

@SethTisue is it possible to run a community build that does not disable `-Xfatal-warnings` to get a feel of how exposed we are to this deprecation?

**Comment by SethTisue** [comment]
> is it possible to run a community build that does not disable `-Xfatal-warnings` to get a feel of how exposed we are to this deprecation?

I think that would give less information than the current setup gives, not more. if we enable `-Xfatal-warnings`, then almost certainly some low-level project will fail and knock out the rest of the build

I think what you want is just to analyze the log of the existing community build and see how often a specific warning we are interested in (such as this one) crops up. I guess we'd want to make sure `-deprecation` is enabled build-wide so that we'd always see the specific form of the warning and not just a "there were 37 deprecation warnings" type message. dbuild supports that.

**Comment by Ichoran** [comment]
@dwijnand - Despite the problems, people who have used inheritance with package objects have already made as much peace as there is to be made with those issues.  So it still seems plausible that the deprecation could wait until there was something with fewer rough edges to migrate to.

@SethTisue - Getting that deprecation log would be great.  I know I use the package object extends pattern a number of times in my code (entirely for implicits, not API exporting), but I can't tell whether I'm an anomaly in that regard.

**Comment by smarter** [comment]
New development: https://github.com/scala/scala/pull/7857

**Pull request merged**